File: app/src/full/java/com/topjohnwu/magisk/asyncs/PatchAPK.java
package com.topjohnwu.magisk.asyncs;

import android.os.AsyncTask;
import android.widget.Toast;

import com.topjohnwu.magisk.BuildConfig;
import com.topjohnwu.magisk.Const;
import com.topjohnwu.magisk.Data;
import com.topjohnwu.magisk.MagiskManager;
import com.topjohnwu.magisk.R;
import com.topjohnwu.magisk.components.Notifications;
import com.topjohnwu.magisk.utils.RootUtils;
import com.topjohnwu.magisk.utils.Utils;
import com.topjohnwu.superuser.ShellUtils;
import com.topjohnwu.utils.JarMap;
import com.topjohnwu.utils.SignAPK;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.CharBuffer;
import java.nio.IntBuffer;
import java.security.SecureRandom;
import java.util.jar.JarEntry;

import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;

public class PatchAPK {

    public static final String LOWERALPHA = "abcdefghijklmnopqrstuvwxyz";
    public static final String UPPERALPHA = LOWERALPHA.toUpperCase();
    public static final String ALPHA = LOWERALPHA + UPPERALPHA;
    public static final String DIGITS = "0123456789";
    public static final String ALPHANUM = ALPHA + DIGITS;
    public static final String ALPHANUMDOTS = ALPHANUM + "............";

    private static String genPackageName(String prefix, int length) {
        StringBuilder builder = new StringBuilder(length);
        builder.append(prefix);
        length -= prefix.length();
        SecureRandom random = new SecureRandom();
        char next, prev = '.';
        for (int i = 0; i < length; ++i) {
            if (prev == '.' || i == length - 1) {
                next = ALPHA.charAt(random.nextInt(ALPHA.length()));
            } else {
                next = ALPHANUMDOTS.charAt(random.nextInt(ALPHANUMDOTS.length()));
            }
            builder.append(next);
            prev = next;
        }
        return builder.toString();
    }

    private static boolean findAndPatch(byte xml[], String a, String b) {
        if (a.length() != b.length())
            return false;
        char[] from = a.toCharArray(), to = b.toCharArray();
        CharBuffer buf = ByteBuffer.wrap(xml).order(ByteOrder.LITTLE_ENDIAN).asCharBuffer();
        int offset = -1;
        for (int i = 0; i < buf.length() - from.length; ++i) {
            boolean match = true;
            for (int j = 0; j < from.length; ++j) {
                if (buf.get(i + j) != from[j]) {
                    match = false;
                    break;
                }
            }
            // Make sure it is null terminated
            if (match && buf.get(i + from.length) == '\0') {
                offset = i;
                break;
            }
        }
        if (offset < 0)
            return false;
        buf.position(offset);
        buf.put(to);
        return true;
    }

    private static boolean findAndPatch(byte xml[], int a, int b) {
        IntBuffer buf = ByteBuffer.wrap(xml).order(ByteOrder.LITTLE_ENDIAN).asIntBuffer();
        int len = xml.length / 4;
        for (int i = 0; i < len; ++i) {
            if (buf.get(i) == a) {
                buf.put(i, b);
                return true;
            }
        }
        return false;
    }

    private static boolean patchAndHide() {
        MagiskManager mm = Data.MM();

        // Generate a new app with random package name
        File repack = new File(mm.getFilesDir(), "patched.apk");
        String pkg = genPackageName("com.", BuildConfig.APPLICATION_ID.length());

        try {
            JarMap apk = new JarMap(mm.getPackageCodePath());
            if (!patch(apk, pkg))
                return false;
            SignAPK.sign(apk, new BufferedOutputStream(new FileOutputStream(repack)));
        } catch (Exception e) {
            return false;
        }

        // Install the application
        repack.setReadable(true, false);
        if (!ShellUtils.fastCmdResult("pm install " + repack))
            return false;;

        mm.mDB.setStrings(Const.Key.SU_MANAGER, pkg);
        Data.exportPrefs();
        RootUtils.rmAndLaunch(BuildConfig.APPLICATION_ID, pkg);

        return true;
    }

    public static boolean patch(JarMap apk, String pkg) {
        try {
            JarEntry je = apk.getJarEntry(Const.ANDROID_MANIFEST);
            byte xml[] = apk.getRawData(je);

            if (!findAndPatch(xml, BuildConfig.APPLICATION_ID, pkg) ||
                    !findAndPatch(xml, BuildConfig.APPLICATION_ID + ".provider", pkg + ".provider") ||
                    !findAndPatch(xml, R.string.app_name, R.string.re_app_name))
                return false;

            // Write in changes
            apk.getOutputStream(je).write(xml);
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }

    public static void hideManager() {
        AsyncTask.THREAD_POOL_EXECUTOR.execute(() -> {
            MagiskManager mm = Data.MM();
            NotificationCompat.Builder progress =
                    Notifications.progress(mm.getString(R.string.hide_manager_title));
            NotificationManagerCompat mgr = NotificationManagerCompat.from(mm);
            mgr.notify(Const.ID.HIDE_MANAGER_NOTIFICATION_ID, progress.build());
            boolean b = patchAndHide();
            mgr.cancel(Const.ID.HIDE_MANAGER_NOTIFICATION_ID);
            if (!b) Utils.toast(R.string.hide_manager_fail_toast, Toast.LENGTH_LONG);
        });
    }
}

File: app/src/main/java/com/topjohnwu/magisk/core/utils/PatchAPK.kt
package com.topjohnwu.magisk.core.utils

import android.content.Context
import android.os.Build.VERSION.SDK_INT
import android.widget.Toast
import com.topjohnwu.magisk.R
import com.topjohnwu.magisk.core.Config
import com.topjohnwu.magisk.core.Const
import com.topjohnwu.magisk.core.Info
import com.topjohnwu.magisk.core.isRunningAsStub
import com.topjohnwu.magisk.data.network.GithubRawServices
import com.topjohnwu.magisk.ktx.get
import com.topjohnwu.magisk.ktx.writeTo
import com.topjohnwu.magisk.utils.Utils
import com.topjohnwu.magisk.view.Notifications
import com.topjohnwu.signing.JarMap
import com.topjohnwu.signing.SignApk
import com.topjohnwu.superuser.Shell
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import timber.log.Timber
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.SecureRandom

object PatchAPK {

    private const val ALPHA = "abcdefghijklmnopqrstuvwxyz"
    private const val ALPHADOTS = "$ALPHA....."

    private const val APP_ID = "com.topjohnwu.magisk"
    private const val APP_NAME = "Magisk Manager"

    private fun genPackageName(prefix: String, length: Int): CharSequence {
        val builder = StringBuilder(length)
        builder.append(prefix)
        val len = length - prefix.length
        val random = SecureRandom()
        var next: Char
        var prev = prefix[prefix.length - 1]
        for (i in 0 until len) {
            next = if (prev == '.' || i == len - 1) {
                ALPHA[random.nextInt(ALPHA.length)]
            } else {
                ALPHADOTS[random.nextInt(ALPHADOTS.length)]
            }
            builder.append(next)
            prev = next
        }
        return builder
    }

    private fun findAndPatch(xml: ByteArray, from: CharSequence, to: CharSequence): Boolean {
        if (to.length > from.length)
            return false
        val buf = ByteBuffer.wrap(xml).order(ByteOrder.LITTLE_ENDIAN).asCharBuffer()
        val offList = mutableListOf<Int>()
        var i = 0
        loop@ while (i < buf.length - from.length) {
            for (j in from.indices) {
                if (buf.get(i + j) != from[j]) {
                    ++i
                    continue@loop
                }
            }
            offList.add(i)
            i += from.length
        }
        if (offList.isEmpty())
            return false

        val toBuf = to.toString().toCharArray().copyOf(from.length)
        for (off in offList) {
            buf.position(off)
            buf.put(toBuf)
        }
        return true
    }

    fun patch(apk: String, out: String, pkg: CharSequence, label: CharSequence): Boolean {
        try {
            val jar = JarMap.open(apk)
            val je = jar.getJarEntry(Const.ANDROID_MANIFEST)
            val xml = jar.getRawData(je)

            if (!findAndPatch(xml, APP_ID, pkg) ||
                !findAndPatch(xml, APP_NAME, label))
                return false

            // Write apk changes
            jar.getOutputStream(je).write(xml)
            val keys = Keygen(get())
            SignApk.sign(keys.cert, keys.key, jar, FileOutputStream(out))
        } catch (e: Exception) {
            Timber.e(e)
            return false
        }

        return true
    }

    private suspend fun patchAndHide(context: Context, label: String): Boolean {
        val dlStub = !isRunningAsStub && SDK_INT >= 28 && Const.Version.atLeast_20_2()
        val src = if (dlStub) {
            val stub = File(context.cacheDir, "stub.apk")
            val svc = get<GithubRawServices>()
            try {
                svc.fetchFile(Info.remote.stub.link).byteStream().use {
                    it.writeTo(stub)
                }
            } catch (e: IOException) {
                Timber.e(e)
                return false
            }
            stub.path
        } else {
            context.packageCodePath
        }

        // Generate a new random package name and signature
        val repack = File(context.cacheDir, "patched.apk")
        val pkg = genPackageName("com.", APP_ID.length)
        Config.keyStoreRaw = ""

        if (!patch(src, repack.path, pkg, label))
            return false

        // Install the application
        if (!Shell.su("adb_pm_install $repack").exec().isSuccess)
            return false

        Config.suManager = pkg.toString()
        Config.export()
        Shell.su("pm uninstall $APP_ID").submit()

        return true
    }

    fun hideManager(context: Context, label: String) {
        val progress = Notifications.progress(context, context.getString(R.string.hide_manager_title))
        Notifications.mgr.notify(Const.ID.HIDE_MANAGER_NOTIFICATION_ID, progress.build())
        GlobalScope.launch {
            val result = withContext(Dispatchers.IO) {
                patchAndHide(context, label)
            }
            if (!result)
                Utils.toast(R.string.hide_manager_fail_toast, Toast.LENGTH_LONG)
            Notifications.mgr.cancel(Const.ID.HIDE_MANAGER_NOTIFICATION_ID)
        }
    }
}
